spec/models/account_suggestions/friends_of_friends_source_spec.rb

```diff
@@ -0,0 +1,82 @@
+    let!(:eve) { Fabricate(:account, discoverable: true, hide_collections: false) }
+    let!(:mallory) { Fabricate(:account, discoverable: false, hide_collections: false) }
+    let!(:eugen) { Fabricate(:account, discoverable: true, hide_collections: false) }
+    let!(:john) { Fabricate(:account, discoverable: true, hide_collections: false) }
+    let!(:jerk) { Fabricate(:account, discoverable: true, hide_collections: false) }
+    let!(:neil) { Fabricate(:account, discoverable: true, hide_collections: false) }
+    let!(:larry) { Fabricate(:account, discoverable: true, hide_collections: false) }
+
+    context 'with follows and blocks' do
+      before do
+        bob.block!(jerk)
+        FollowRecommendationMute.create!(account: bob, target_account: neil)
+
+        # bob follows eugen, alice and larry
+        [eugen, alice, larry].each { |account| bob.follow!(account) }
+
+        # alice follows eve and mallory
+        [john, mallory].each { |account| alice.follow!(account) }
+
+        # eugen follows eve, john, jerk, larry and neil
+        [eve, mallory, jerk, larry, neil].each { |account| eugen.follow!(account) }
+      end
+
+      it 'returns eligible accounts', :aggregate_failures do
+        results = subject.get(bob)
+
+        # eve is returned through eugen
+        expect(results).to include([eve.id, :friends_of_friends])
+
+        # john is not reachable because alice hides who she follows
+        expect(results).to_not include([john.id, :friends_of_friends])
+
+        # mallory is not discoverable
+        expect(results).to_not include([mallory.id, :friends_of_friends])
+
+        # larry is not included because he's followed already
+        expect(results).to_not include([larry.id, :friends_of_friends])
+
+        # jerk is blocked
+        expect(results).to_not include([jerk.id, :friends_of_friends])
+
+        # the suggestion for neil has already been rejected
+        expect(results).to_not include([neil.id, :friends_of_friends])
+      end
+    end
+
+    context 'with deterministic order' do
+      before do
+        # bob follows eve and mallory
+        [eve, mallory].each { |account| bob.follow!(account) }
+
+        # eve follows eugen, john, and jerk
+        [jerk, eugen, john].each { |account| eve.follow!(account) }
+
+        # mallory follows eugen, john, and neil
+        [neil, eugen, john].each { |account| mallory.follow!(account) }
+
+        john.follow!(eugen)
+        john.follow!(neil)
+      end
+
+      it 'returns eligible accounts in the expected order' do
+        expect(subject.get(bob)).to eq [
+          [eugen.id, :friends_of_friends], # followed by 2 friends, 3 followers total
+          [john.id, :friends_of_friends], # followed by 2 friends, 2 followers total
+          [neil.id, :friends_of_friends], # followed by 1 friend, 2 followers total
+          [jerk.id, :friends_of_friends], # followed by 1 friend, 1 follower total
```

Comment:
I think there may be an issue in the query here.

We have a bunch of recent failures which look intermittent, ie - https://github.com/mastodon/mastodon/actions/runs/8190281546/job/22397018506?pr=29517

I had trouble replicating this locally, but I did a little debugging and it looks like the "followed by" value here is just `1` for every record. I think what's happening in the spec is that each record has a `1, N` value its being sorted on. This means the second value will sort the `eugen` and `jerk` accounts to be first/last correctly, and we'd expect it to be a coin flip for the middle two since they would both be `1,2`.

On CI, it does seem like a coin flip, but locally it always passes. I was able to get consistent local failures by leaving the class code and spec assertions alone - but just changing the creation order (ie, move `neil` to be above `john`) which generates consistent failures. This difference might come down to different postgres versions or default sort order or something (I assume its using lowest-account-ID as local default).

I haven't looked in detail at the recent changes to the query yet ... my first guess is that something which used to be counting joined rows has lost that join and is now just counting the already-grouped rows (and thus getting `1` for every row).

Assuming this sounds plausible - it may be worth doing some refactor here so that we can assert things about the query results themselves before they are wrapped in the pluck/map that the source class uses ... maybe pull out the bulk of the method before the pluck/map to another public method?